lib/seattleflu/db/cli/command/etl/consensus_genome.py

```diff
@@ -0,0 +1,363 @@
+    LOG.debug(f"Looking up sample «{identifier}»")
+
+    sample = db.fetch_row("""
+        select sample_id as id, identifier, encounter_id
+          from warehouse.sample
+         where identifier = %s or
+               collection_identifier = %s
+        """, (identifier,identifier,))
+
+    if not sample:
+        LOG.error(f"No sample with identifier «{identifier}» found")
+        return None
+
+    LOG.info(f"Found sample {sample.id} «{sample.identifier}»")
+    return sample
+
+
+def find_or_create_sequence_read_set(db: DatabaseSession, document: dict, sample: Any) -> Any:
+    """
+    Find sequence read set given a *sample* and consensus genome record
+    *document*, inserting the sequence read set if it does not exist.
+    Return the sequence read set.
+    """
+    urls = document['metadata']['URLs']
+    LOG.debug(dedent(f"""
+    Looking up sequence read set with sample ID «{sample.id}» and URLs {urls}
+    """))
+
+    sequence_read_set = db.fetch_row("""
+        select sequence_read_set_id as id, sample_id, urls
+          from warehouse.sequence_read_set
+         where sample_id = %s
+          and urls @> %s
+          and %s @> urls
+        """, (sample.id, urls, urls))
+
+    if sequence_read_set:
+        LOG.info(f"Found sequence read set {sequence_read_set.id}")
+    else:
+        LOG.debug(dedent(f"""
+        Sequence read set not found for sample id «{sample.id}» and URLs {urls}
+        """))
+
+        data = {
+            "sample_id": sample.id,
+            "urls": urls,
+        }
+
+        sequence_read_set = db.fetch_row("""
+            insert into warehouse.sequence_read_set (sample_id, urls)
+                values (%(sample_id)s, %(urls)s)
+            returning sequence_read_set_id as id
+            """, data)
+
+        LOG.info(f"Created sequence read set {sequence_read_set.id}")
+
+    return sequence_read_set
+
+
+def mark_received(db, sequence_read_set_id: int, entry: {}) -> None:
+    """
+    This function is a workaround to the order-specific unique constraint for
+    arrays in Postgres. It searches for an existing *sequence_read_set* by its
+    ID and updates the details column with the given *entry* details.
+    """
+    LOG.debug(f"Marking sequence read set data {sequence_read_set_id} as received")
+
+    data = {
+        "sequence_read_set_id": sequence_read_set_id,
+        "log_entry": Json(entry),
+    }
+
+    # Postgres allows array columns to be defined as unique, but the ordering of
+    # the arrays must be the same for the arrays to match. We are only
+    # interested in matching on array content, not array ordering. This prevents
+    # us from being able to use the built-in unique constraint on URLs, thereby
+    # preventing `ON CONFLICT` updates to a table when URLs is the only unique
+    # column in the table.
+    with db.cursor() as cursor:
+        cursor.execute("""
+            update warehouse.sequence_read_set
+              set details = %(log_entry)s
```

Comment:
Might want to change this to `set details = coalesce(details, '{}') || %(log_entry)s` so that the previous details do not get overwritten.

Reply:
Assuming the log entry has the same keys, because the coalescing happens at the top level of the Json object, won't the items with the same keys get overwritten anyway?

Reply:
Hmm that's true, then maybe we want to make `sequence_read_set.details` a nested json somehow?

Reply:
What are you interested in tracking in the details column? I think if I better understand your intended use case, I can design something that makes more sense.

Reply:
I updated the `details` column for sequence read set to your suggestion. Now that we have different top-level keys depending on the target organism, I changed the SQL to use coalesce.